datafusion/expr/src/udf.rs

```diff
@@ -18,11 +18,30 @@
 //! Udf module contains foundational types that are used to represent UDFs in DataFusion.
 
 use crate::{Expr, ReturnTypeFunction, ScalarFunctionImplementation, Signature};
+use arrow::array::ArrayRef;
+use datafusion_common::Result;
 use std::fmt;
 use std::fmt::Debug;
 use std::fmt::Formatter;
 use std::sync::Arc;
 
+pub trait ScalarFunctionDef: Sync + Send + std::fmt::Debug {
+    // TODO: support alias
+    fn name(&self) -> &str;
+
+    fn signature(&self) -> Signature;
+
+    // TODO: ReturnTypeFunction -> a ENUM
+    //     most function's return type is either the same as 1st arg or a fixed type
+    fn return_type(&self) -> ReturnTypeFunction;
```

Comment:
I think the current API requires return_type to be passed the specific argument types

Also, https://github.com/apache/arrow-datafusion/discussions/7657 suggests there is additional room for improvement

Reply:
I think we might mark this API as experimental for now, it can have frequent changes to this API, other unexpected cases like #7657 might show up when porting some trickier functions.

Reply:
I like that idea -- 

maybe the implementation plan sequence could be:
1. Introduce `ScalarFunctionDef` 
2. `impl ScalarFunctionDef` for `ScalarUDF` and `BuiltInFunction`
3. Update the core DataFusion code to only use `ScalarFunctionDef`
4. Port all BuiltInFunction to `ScalarFunctionDef`
5. Deprecate `ScalarFunctionDef` 
6. Remove `BuiltinFunctionDef`

Reply:
Sounds great 👍🏼 My plan is
1st PR: `impl ScalarFunctionDef for BuiltinScalarFunction` and replace the execution code with the new interface. This step make sure the new interface can cover all functionalities, and determine how the interface looks like
2nd PR: replace current UDF impl with new interface
3nd and after: start porting existing functions